test/gtest/ucp/test_ucp_reconfigure.cc

```diff
@@ -0,0 +1,272 @@
+                                    ucp_ep->conn_sn, UCS_CONN_MATCH_QUEUE_EXP));
+
+    ASSERT_UCS_OK(ucp_wireup_send_request(ucp_ep));
+    UCS_ASYNC_UNBLOCK(&worker()->async);
+
+    store_config();
+}
+
+bool test_ucp_reconfigure::entity::has_matching_lane(ucp_ep_h ep,
+                                                     ucp_lane_index_t lane_idx,
+                                                     const entity &other) const
+{
+    const auto lane     = &ucp_ep_config(ep)->key.lanes[lane_idx];
+    const auto resource = &ucph()->tl_rscs[ucp_ep_get_rsc_index(ep, lane_idx)];
+    auto addr           = other.get_address(true);
+    const ucp_address_entry_t *ae;
+
+    ucs_carray_for_each(ae, addr->second.address_list,
+                        addr->second.address_count) {
+        if ((resource->tl_name_csum == ae->tl_name_csum) &&
+            ucp_wireup_is_lane_connected(ep, lane_idx, ae)) {
+            EXPECT_EQ(ae->sys_dev, lane->dst_sys_dev);
+            EXPECT_EQ(ae->md_index, lane->dst_md_index);
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void test_ucp_reconfigure::entity::verify(const entity &other) const
+{
+    unsigned reused                  = 0;
+    const ucp_lane_index_t num_lanes = ucp_ep_num_lanes(ep());
+
+    EXPECT_EQ(num_lanes, ucp_ep_num_lanes(other.ep()));
+
+    for (ucp_lane_index_t lane = 0; lane < num_lanes; ++lane) {
+        /* Verify local and remote lanes are identical */
+        EXPECT_TRUE(has_matching_lane(ep(), lane, other));
+
+        /* Verify correct number of reused lanes is configured */
+        auto uct_ep = ucp_ep_get_lane(ep(), lane);
+        auto it     = std::find(m_uct_eps.begin(), m_uct_eps.end(), uct_ep);
+        reused     += (it != m_uct_eps.end());
+    }
+
+    EXPECT_EQ(reused, is_reconfigured() ? 0 : num_lanes);
+}
+
+test_ucp_reconfigure::address_pair_p
+test_ucp_reconfigure::entity::get_address(bool ep_only) const
+{
+    const ucp_tl_bitmap_t tl_bitmap = ep_only ? ep_tl_bitmap() :
+                                                ucp_tl_bitmap_max;
+    size_t addr_len;
+
+    address_pair_p address(new address_pair_t(nullptr, {0}),
+                           [](address_pair_t *pair) {
+                               ucs_free(pair->first);
+                               ucs_free(pair->second.address_list);
+                               delete pair;
+                           });
+
+    unsigned flags = ucp_worker_default_address_pack_flags(worker());
+    ASSERT_UCS_OK(ucp_address_pack(worker(), ep(), &tl_bitmap, flags,
+                                   ucph()->config.ext.worker_addr_version, NULL,
+                                   UINT_MAX, &addr_len, &address->first));
+
+    ASSERT_UCS_OK(ucp_address_unpack(worker(), address->first, flags,
+                                     &address->second));
+    return address;
+}
+
+UCS_TEST_P(test_ucp_reconfigure, basic)
+{
+    create_entity();
+    create_entity();
+    send_recv();

```

Comment:
Maybe the `connect` should be in something like `create_entities_and_connect`, but definitely separated from `send_recv`.

To support future more sophisticated scenarios.